events/handler.go

```diff
@@ -0,0 +1,196 @@
+                watchers:  make(map[*Request]chan *Event),
+        }
+}
+
+// returns a pointer to an initialized Request object
+func NewRequest() *Request {
+        return &Request{
+                EventType: map[EventType]bool{},
+        }
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Len() int {
+        return len(e)
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Swap(i, j int) {
+        e[i], e[j] = e[j], e[i]
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Less(i, j int) bool {
+        return e[i].Timestamp.Before(e[j].Timestamp)
+}
+
+// sorts and returns up to the last MaxEventsReturned chronological elements
+func getMaxEventsReturned(request *Request, eSlice EventSlice) EventSlice {
+        sort.Sort(eSlice)
+        n := request.MaxEventsReturned
+        if n >= eSlice.Len() || n <= 0 {
+                return eSlice
+        }
+        return eSlice[eSlice.Len()-n:]
+}
+
+// determines if an event occurs within the time set in the request object and is the right type
+func checkIfEventSatisfiesRequest(request *Request, event *Event) bool {
+        startTime := request.StartTime
+        endTime := request.EndTime
+        eventTime := event.Timestamp
+        if !startTime.IsZero() {
+                if startTime.After(eventTime) {
+                        return false
+                }
+        }
+        if !endTime.IsZero() {
+                if endTime.Before(eventTime) {
+                        return false
+                }
+        }
+        if request.EventType[event.EventType] != true {
+                return false
+        }
+        return true
+}
+
+// method of Events object that screens Event objects found in the eventlist
+// attribute and if they fit the parameters passed by the Request object,
+// adds it to a slice of *Event objects that is returned. If both MaxEventsReturned
+// and StartTime/EndTime are specified in the request object, then only
+// up to the most recent MaxEventsReturned events in that time range are returned.
+func (self *events) GetEvents(request *Request) (EventSlice, error) {
+        returnEventList := EventSlice{}
+        for _, e := range self.eventlist {
+                if checkIfEventSatisfiesRequest(request, e) {
+                        returnEventList = append(returnEventList, e)
+                }
+        }
+        sort.Sort(returnEventList)
+        returnEventList = getMaxEventsReturned(request, returnEventList)
+        return returnEventList, nil
+}
+
+// method of Events object that maintains an *Event channel passed by the user.
+// When an event occurs that satisfies the parameters in the passed Request
+// object it is fed to the channel
+func (self *events) WatchEvents(outChannel chan *Event, request *Request) error {
+        if self.watchers[request] == nil {
```

Comment:
Unfortunately we may not be able to make this a map since we may have more than one user with the same request. Imagine two calls to Watch() that only want OOM events, they should be handled as separate things. Maybe a slice with a request and a channel?